docs/concepts/tools.mdx

```diff
@@ -253,6 +253,82 @@ writer1 = Agent(
     #...
 ```
 
+## Tool Collection
+
+Tool Collection allow users to load a collection of CrewAI tools from various sources.
+For example CrewAI Tool Collection can leverage tools from a [Model Context Protocol (MCP)](https://modelcontextprotocol.io/introduction) server.
+
+### MCP Tools
+
+### what is MCP?
+
+MCP is an open source protocol which allow anyone to create servers to define tools among other things.
+Since its release in november 2024 a large amount of MCP server have been created by the community covering a large panel of tools that can be leveraged by your agents.
+Some websites index available MCP server (600+) like [glama.ai](https://glama.ai/mcp/servers) or [smithery.ai](https://smithery.ai/).
+
+Today MCP servers leverage 2 possible transport protocol to communicate with a client: stdio with a local subprocess or SSE with a remote server.
+Both transport protocol are supported with CrewAI Tool Collection capabilities.
+
+<Warning>
+  When using MCP servers over stdio transport protocol, be extremely cautious and verify the authenticity and trustworthiness of the server.
+  The stdio protocol allows the MCP server to execute programs on your machine with your permissions.
+  Only use MCP servers from trusted sources and carefully review their code before running them.
+</Warning>
+
+### How to leverage MCP server tools with CrewAI?
+
+MCP servers usually provides several tools to load several tools from an MCP server you can leverage `ToolCollection.from_mcp`.
+
+Before being able to use `ToolCollection.from_mcp` you need to install the extra `mcp` depedencies either via uv:
+
+```bash
+uv sync --extra mcp
+```
+
+or via pip
+
+```bash
+pip install crewai[mcp]
+# or
+uv pip install crewai[mcp]
+```
+
+Here is an example where we leverage the pubmedmcp MCP server to load a tool that allow our agent to search abstract of medical papers from pubmed:
+
+```python Code
+from crewai import Agent, Task, Crew
+from crewai.tools import ToolCollection
+from mcp import StdioServerParameters
+
+server_parameters = StdioServerParameters(
+    command="uv",
+    args=["--quiet", "pubmedmcp@0.1.3"],
+    env={"UV_PYTHON": "3.12", **os.environ},
+)
+
+with ToolCollection.from_mcp(server_parameters) as tool_collection:
+    agent = Agent(
+        role="Research Agent",
+        goal="Find studies about hangover",
+        backstory="You help find studies about hangover",
+        verbose=True,
+        tools=[tool_collection.tools[0]],
+    )
+    task = Task(
+        description="Find studies about hangover",
+        agent=agent,
+        expected_output="A list of studies about hangover",
+    )
+    crew = Crew(agents=[agent], tasks=[task], verbose=True)
+    result = crew.kickoff()
```

Comment:
is there a way to not have a wrapper like this and instead have it setup within tools. Ideal implementation would look like this for me, and something that would live on the tools repo vs the core lib:

WDYT of this:
```python
mcp_tool = MCPServerToolAdapter(server_parameters={...})
agent = Agent(
        role="Research Agent",
        goal="Find studies about hangover",
        backstory="You help find studies about hangover",
        verbose=True,
        tools=[mcp_tool],
    )
```

I started this: but agents struggling to pass the arguments properly. If you can help bring your implementation over 🙏🏼 
https://github.com/crewAIInc/crewAI-tools/tree/feat/mcp-server-tools-adapter

Reply:
I don't see your implementation there did you push it? 
It's definitely possible to make it more as above with a few caveats: here I suggested to use a context manager because unlike other kind of tool that I have seen in the repo attached, we need to manage the lifecycle of the mcp server so when you start the context manager you actually spawn the mcp server in the background in a subprocess. When you reach the end of your agentic pipeline you want to cleanly kill the subprocess / mcp server so contextmanager really fit that idea here. That being said, it's also possible to do it as you did but then users would need to not forget to somehow close / kill the mcp server process after the agent has run.

Reply:
we can spin up and then close when the tool call is complete. would that work ?

Reply:
wondering if there is a more elegant DX wbhere we manage that.